tests/dotnet/GCResurrectApp/Main.cs

```diff
@@ -0,0 +1,63 @@
+using System;
+using Foundation;
+using AppKit;
+
+public class Program
+{
+	[System.Runtime.CompilerServices.MethodImpl(System.Runtime.CompilerServices.MethodImplOptions.NoInlining)]
+	public static IntPtr CreateNativeObject()
+	{
+		var foo = new NSString("foo");
+		foo.DangerousRetain();
+		return foo.Handle;
+	}
+
+	public static void PumpLoop()
+	{
+		using (new NSAutoreleasePool())
+		{
+			NSEvent? evnt;
+			while ((evnt = NSApplication.SharedApplication.NextEvent((NSEventMask)NSEventMask.AnyEvent, NSDate.DistantPast, NSRunLoopMode.Default, true)) != null)
+			{
+				NSApplication.SharedApplication.SendEvent(evnt);
+			}
+		}
+	}
+
+	public static bool NSObjectFinalizeTest()
+	{
+		// Pump run loop to ensure that NSObject_Disposer has processed everything
+		PumpLoop();
```

Comment:
It would not work reliably if the unit tests are run in parallel. it's also hard to guarantee correctness of the test if the test runner is in control of the UI thread loop.

Reply:
🤔 - Is there an attribute that lets us say "run this not in parallel"?

Reply:
In xUnit/nUnit this is quite hard to achieve but not impossible. They offer different levels of parallelization. Unfortunately the test needs a significant degree of precision over what happens and the test runner may inadvertently interfer with that. The whole NSObject_Disposer logic triggers based on the length of the queue of objects to be finalized and the test depends precisely on when that triggers. If any other caller allocated something derived from NSObject (and made it eligible for GC) it could break.

The whole reason why I put it into separate project is to ensure that it is single threaded and I have total precision about what happens in the process. It is possible to relax it a bit but it also makes it significantly harder to reason about its correctness.

Reply:
> It would not work reliably if the unit tests are run in parallel [within the same process].

We don't run tests in parallel in this particular test project (they're mostly UI tests, so everything has to happen on the UI thread anyway).

> The whole reason why I put it into separate project is to ensure that it is single threaded and I have total precision about what happens in the process.

That's a good point, but you could also argue the other way: the test should succeed no matter what the process does in other threads... (not that I'd ever want to track down any such problems!)

> It is possible to relax it a bit but it also makes it significantly harder to reason about its correctness.

Given that the tests are already executed serially, what other correctness issues do you see?

Reply:
If the guarantee is that it runs on UI thread and the test runner is single-threaded then it should be doable. I'll give it a try.

> the test should succeed no matter what the process does in other threads...

I am basically decomposing a multi-threaded race condition into its individual elements which I force to run in serialized manner on UI thread to insert deterministic order into it. If I allowed multiple threads to run I would be removing significant determinism factor from the test (ie. other objects getting collected and finalized and trying to trigger something on the UI thread).

Reply:
I converted the test but...

> Another advantage is that the test would be executed on all our platforms without any other work.

Unfortunately the test won't work on other platforms for two different reasons:
- Mono has conservative garbage collection and it's currently not possible to precisely trigger collection of an object. I could move the object creation into different thread which would partly avoid problem with conservative stack scanning but some other stack could accidentally have the same numerical value and the test would still fail.
- The test requires pumping the message loop which uses different APIs on UIKit and AppKit. Since it would not work anyway I left only the AppKit version.

I may have missed some guard for the test that would prevent it from running on legacy Xamarin.Mac+Mono.

Reply:
> Mono has conservative garbage collection and it's currently not possible to precisely trigger collection of an object. I could move the object creation into different thread which would partly avoid problem with conservative stack scanning but some other stack could accidentally have the same numerical value and the test would still fail (or rather pass and not test for the broken scenario).

This means that the test may succeed sometimes when it shouldn't - but there's still value in it because it should never fail when it shouldn't (in other words: a failing test with Mono is always a real failure, which is valuable even if the a passing test doesn't necessarily mean nothing is broken).